pkg/virtctl/usbredir/client.go

```diff
@@ -0,0 +1,188 @@
+		ClientConnect: clientConnect,
+	}
+}
+
+func (k *Client) WithRemoteVMIStream(usbredirStream kubecli.StreamInterface) *Client {
+	k.stream = make(chan error)
+
+	go func() {
+		defer k.outputWriter.Close()
+		k.stream <- usbredirStream.Stream(
+			kvcorev1.StreamOptions{
+				In:  k.inputReader,
+				Out: k.outputWriter,
+			},
+		)
+	}()
+
+	return k
+}
+
+func (k *Client) WithLocalTCPClient(address string) *Client {
+	lnAddr, err := net.ResolveTCPAddr("tcp", address)
+	if err != nil {
+		log.Log.Errorf("Can't resolve the address: %s", err.Error())
+		return nil
+	}
+
+	// The local tcp server is used to proxy between remote websocket and local USB
+	k.listener, err = net.ListenTCP("tcp", lnAddr)
+	if err != nil {
+		log.Log.Errorf("Can't listen on unix socket: %s", err.Error())
+		return nil
+	}
+
+	return k
+}
+
+func (k *Client) ConnectRemote() {
+	// forward data to/from websocket after usbredir client connects.
+	k.done = make(chan struct{}, 1)
+	k.remote = make(chan error)
+	go func() {
+		defer k.inputWriter.Close()
+		start := time.Now()
+
+		usbredirConn, err := k.listener.Accept()
+		if err != nil {
+			log.Log.V(2).Infof("Failed to accept connection: %s", err.Error())
+			k.remote <- err
+			return
+		}
+		defer usbredirConn.Close()
+
+		log.Log.V(2).Infof("Connected to %s at %v", usbredirClient, time.Now().Sub(start))
+
+		stream := make(chan error)
+		// write to local usbredir from pipeOutReader
+		go func() {
+			_, err := io.Copy(usbredirConn, k.outputReader)
+			stream <- err
+		}()
+
+		// read from local usbredir towards pipeInWriter
+		go func() {
+			_, err := io.Copy(k.inputWriter, usbredirConn)
+			stream <- err
+		}()
+
+		select {
+		case <-k.done: // Wait for local usbredir to complete
+		case err = <-stream: // Wait for remote connection to close
+			if err == nil {
+				// Remote connection closed, report this as error
+				err = fmt.Errorf("Remote connection has closed.")
+			}
+		}
+
+		// Wait for local usbredir to complete
+		k.remote <- err
+	}()
+}
+
+func clientConnect(device, address string) error {
+	bin := usbredirClient
+	args := []string{}
+	args = append(args, "--device", device, "--to", address)
+
+	log.Log.Infof("port_arg: '%s'", address)
+	log.Log.Infof("args: '%v'", args)
+	log.Log.Infof("Executing commandline: '%s %v'", bin, args)
+
+	command := exec.Command(bin, args...)
+	output, err := command.CombinedOutput()
+	if err != nil {
+		log.Log.Errorf("Failed to execute %v due %v, output: %v", bin, err, string(output))
+	} else {
+		log.Log.V(2).Infof("%v output: %v", bin, string(output))
+	}
+	return err
+}
+
+func (k *Client) ConnectLocal(device string) {
+	// execute local usbredir binary
+	address := k.listener.Addr().String()
+	k.local = make(chan error)
+	go func() {
+		defer close(k.done)
+		k.local <- k.ClientConnect(device, address)
+	}()
+}
+
+func (k *Client) Run() error {
+	var err error
+
+	interrupt := make(chan os.Signal, 1)
+	signal.Notify(interrupt, os.Interrupt)
```

Comment:
I would say you don't want to handle interrupts here as you could reuse the client in different context. I think it would be cleaner to leave it up to user which would than be able to call `Stop` or handle it otherwise.

Reply:
Perhaps I'll add a WithSignalInterrupt() as both virtctl and the test might use this and we can opt-out if not.

Reply:
I moved that handling to `usbredir.go` instead. I don't expect we using `client.go` anywhere else _but_ the tests which I did not bring the signal handler in.